### PR TITLE
fixing url check for mobile page

### LINKED ISSUE #1322: Missing h1 article heading if customMainPage is not the exactly articleId
For example, with the `wikipedia_en_ray_charles_maxi_2020-12.zim`, none of the articles have the h1 header at the top of the page (see screenshot of article from December ZIM [top] compared with article for November ZIM [bottom]).

With `wikipedia_en_computer_maxi_2020-12.zim` some articles have h1 and others don't.

**Ray Charles December 2020:**

image

**Ray Charles November 2020:**

image

### PR DESCRIPTION
**Pull request by MananJethwani**
this temporarily resolves #1322

### CODE REVIEW

**Review by kelson42 — CHANGES_REQUESTED** [review]
Thank you for proposing quickly a fix, but would that not be healthier to just check/fix things a the sanitize step?

File: src/mwoffliner.lib.ts [hunk 1]
```diff
@@ -197,7 +197,7 @@ async function execute(argv: any) {
   }
   const mainPage = customMainPage || (articleList ? '' : mwMetaData.mainPage);
 
-  await downloader.checkCapabilities(mainPage);
+  await downloader.checkCapabilities(mainPage, mwMetaData.mainPage);
```
Comment (kelson42) [inline src/mwoffliner.lib.ts:200]: Why adding a new argument? I thought we agre we should always test with `mwMetaData.mainPage`?

File: src/sanitize-argument.ts [hunk 2]
```diff
@@ -107,6 +122,30 @@ export function sanitize_redis(argv:any)
...
 }
 
+export async function checkApiAvailabilty(url: string): Promise<boolean>{
+  try {
+    const resp = await axios.get(url, { maxRedirects: 0 });
+    return resp.status === 200 && !resp.headers['mediawiki-api-error'];
+  } catch (err) {
+    return false;
+  }
+}
```
Comment (kelson42) [inline src/sanitize-argument.ts:132]: Full code duplicate, this should never happen.

File: src/sanitize-argument.ts [hunk 3]
```diff
@@ -107,6 +122,30 @@ export function sanitize_redis(argv:any)
...
+  const veApiUrl = new URL(`${apiUrl.href}action=visualeditor&mobileformat=html&format=json&paction=parse&page=`);
+  const desktopRestApiUrl = new URL(ensureTrailingChar(new URL(restApiPath ?? 'api/rest_v1/page/html', baseUrl.href).toString(), '/'));
+  const baseUrlForMainPage = checkApiAvailabilty(desktopRestApiUrl.href + encodeURIComponent(customMainPage)) ? desktopRestApiUrl.href :
+                             checkApiAvailabilty(veApiUrl.href) ? veApiUrl.href :
+                             undefined;
+  if (!checkApiAvailabilty(baseUrlForMainPage + encodeURIComponent(customMainPage)) && noLocalParserFallback) {
+    throw new Error(`Failed to verify status code 200 on custom main page`)
+  }
+  return customMainPage.replace(/ /g,'_')
+}
```
Comment (kelson42) [inline src/sanitize-argument.ts:147]: In  #1318, I just ask to test if the `customMainPage` article exists at all. I'm sure we can do a lot simpler. Just checking if `baseUrl + customMainPage` gives HTTP 200 should be sufficient, shouldn't it!?

**Review by kelson42 — CHANGES_REQUESTED** [review]

File: src/sanitize-argument.ts (hunk 3 above)
Comment (MananJethwani) [reply src/sanitize-argument.ts:147]: ok

File: src/mwoffliner.lib.ts (hunk 1 above)
Comment (MananJethwani) [reply src/mwoffliner.lib.ts:200]: that info is not available in downloader class need to pass that here

Comment (MananJethwani) [reply src/mwoffliner.lib.ts:200]: ok I get it sorry

File: src/sanitize-argument.ts (hunk 2 above)
Comment (kelson42) [reply src/sanitize-argument.ts:132]: I don't think you need that anyway, you just need to make a simple get request on a URL and check if code is HTTP 200. You can do that directly with Axios.

File: src/sanitize-argument.ts [hunk 4]
```diff
@@ -107,6 +120,19 @@ export function sanitize_redis(argv:any)
...
+    if(res.status !== 200) {
+      throw new Error('Failed to verify status code 200 on custom main page')
+    }
+  })
+  .catch((err) => {
+    throw err;
+  });
+  return customMainPage.replace(/ /g,'_');
+}
+
```
Comment (kelson42) [inline src/sanitize-argument.ts:135]: This function should be remove as most of the line is the exception mgmt, which is duplicate with lin 61.

We should not have this `wiki` here, this is a default value which should be handled only in the Mediawiki module.

`replace(/ /g,'_')` should not be done twice. Only once.

**Review by kelson42 — CHANGES_REQUESTED** [review]

File: src/sanitize-argument.ts [hunk 5]
```diff
@@ -55,6 +56,20 @@ export async function sanitize_all(argv: any) {
     throw err;
   });
 
+  // sanitize Custom Main Page
+  if (argv.customMainPage) {
+    await axios.get(`${mwUrl}/wiki/${argv.customMainPage.replace(/ /g,'_')}`)
```
Comment (kelson42) [inline src/sanitize-argument.ts:61]: * we should keep the space replacement
* you shoudl not hardcode the path `wiki`, you should use whatever hase been given in Mediawiki.

**Review by kelson42 — CHANGES_REQUESTED** [review]

File: src/sanitize-argument.ts [hunk 6]
```diff
@@ -55,6 +56,20 @@ export async function sanitize_all(argv: any) {
     throw err;
   });
 
+  // sanitize Custom Main Page
+  if (argv.customMainPage) {
+    argv.customMainPage = argv.customMainPage.replace(/ /g,'_');
+    await axios.get(`${mwUrl}/${mwWikiPath || defaultWikiPath}${argv.customMainPage.replace(/ /g,'_')}`)
```
Comment (kelson42) [inline src/sanitize-argument.ts:62]: Underscore replacement is not necessary

File: src/util/const.ts [hunk 7]
```diff
@@ -10,4 +10,5 @@ export const DB_ERROR = 'internal_api_error_DBQueryError';
 export const WEAK_ETAG_REGEX = /^(W\/)/;
 export const WEBP_CANDIDATE_IMAGE_URL_REGEX = /(\.jpeg|\.png|\.jpg)(\?.*)?$/i;
 export const BITMAP_IMAGE_MIME_REGEX = /^image+[/-\w.]+(jpeg|png|gif)$/;
-export const WEBP_CANDIDATE_IMAGE_MIME_TYPE = /image+[/]+(jpeg|png)/;
\ No newline at end of file
+export const WEBP_CANDIDATE_IMAGE_MIME_TYPE = /image+[/]+(jpeg|png)/;
+export const defaultWikiPath = 'wiki/'
```
Comment (kelson42) [inline src/util/const.ts:14]: Constant name should be uppercase

File: src/sanitize-argument.ts (hunk 6 above)
Comment (kelson42) [inline src/sanitize-argument.ts:62]: Redirects should be allowed in a ios request

File: src/sanitize-argument.ts [hunk 8]
```diff
@@ -55,6 +56,20 @@ export async function sanitize_all(argv: any) {
...
+    .then((res) => {
+      if(res.status !== 200) {
+        throw new Error(`customMainPage doesn't return 200 status code`);
```
Comment (kelson42) [inline src/sanitize-argument.ts:65]: The error should inclide the tested url.

File: src/sanitize-argument.ts [hunk 9]
```diff
@@ -55,6 +56,20 @@ export async function sanitize_all(argv: any) {
...
+      }
+    })
+    .catch((err) => {
+      throw err;
```
Comment (kelson42) [inline src/sanitize-argument.ts:69]: Why we have a ctsch() here, considering that we have one at a higher level araiund sanitizeAll()

**Review by kelson42 — CHANGES_REQUESTED** [review]

File: src/sanitize-argument.ts (hunk 9 above)
Comment (MananJethwani) [reply src/sanitize-argument.ts:69]: yeah can be removed

File: src/sanitize-argument.ts (hunk 6 above)
Comment (MananJethwani) [reply src/sanitize-argument.ts:62]: @kelson42 redirects are allowed by default in axios

**Review by kelson42 — APPROVED** [review]

**Pull request merged**